core/services/ocr2/plugins/ocr2keeper/evm/log_provider.go

```diff
@@ -0,0 +1,483 @@
+		logger:            logger,
+		logPoller:         logPoller,
+		registryAddress:   registryAddress,
+		lookbackBlocks:    lookbackBlocks,
+		registry:          contract,
+		client:            client,
+		packer:            NewEvmRegistryPackerV2_1(abi),
+		txCheckBlockCache: pluginutils.NewCache[string](time.Hour),
+		cacheCleaner:      pluginutils.NewIntervalCacheCleaner[string](time.Minute),
+	}, nil
+}
+
+func (c *LogProvider) Name() string {
+	return c.logger.Name()
+}
+
+func (c *LogProvider) Start(ctx context.Context) error {
+	return c.sync.StartOnce("AutomationLogProvider", func() error {
+		c.mu.Lock()
+		defer c.mu.Unlock()
+
+		go c.cacheCleaner.Run(c.txCheckBlockCache)
+		c.runState = 1
+		return nil
+	})
+}
+
+func (c *LogProvider) Close() error {
+	return c.sync.StopOnce("AutomationRegistry", func() error {
+		c.mu.Lock()
+		defer c.mu.Unlock()
+
+		c.cacheCleaner.Stop()
+		c.runState = 0
+		c.runError = nil
+		return nil
+	})
+}
+
+func (c *LogProvider) Ready() error {
+	c.mu.RLock()
+	defer c.mu.RUnlock()
+
+	if c.runState == 1 {
+		return nil
+	}
+	return c.sync.Ready()
+}
+
+func (c *LogProvider) HealthReport() map[string]error {
+	c.mu.RLock()
+	defer c.mu.RUnlock()
+
+	if c.runState > 1 {
+		c.sync.SvcErrBuffer.Append(fmt.Errorf("failed run state: %w", c.runError))
+	}
+	return map[string]error{c.Name(): c.sync.Healthy()}
+}
+
+func (c *LogProvider) PerformLogs(ctx context.Context) ([]ocr2keepers.PerformLog, error) {
+	end, err := c.logPoller.LatestBlock(pg.WithParentCtx(ctx))
+	if err != nil {
+		return nil, fmt.Errorf("%w: failed to get latest block from log poller", err)
+	}
+
+	// always check the last lookback number of blocks and rebroadcast
+	// this allows the plugin to make decisions based on event confirmations
+	logs, err := c.logPoller.LogsWithSigs(
+		end-c.lookbackBlocks,
+		end,
+		[]common.Hash{
+			iregistry21.IKeeperRegistryMasterUpkeepPerformed{}.Topic(),
+		},
+		c.registryAddress,
+		pg.WithParentCtx(ctx),
+	)
+	if err != nil {
+		return nil, fmt.Errorf("%w: failed to collect logs from log poller", err)
+	}
+
+	performed, err := c.unmarshalPerformLogs(logs)
+	if err != nil {
+		return nil, fmt.Errorf("%w: failed to unmarshal logs", err)
+	}
+
+	vals := []ocr2keepers.PerformLog{}
+	for _, p := range performed {
+		upkeepId := ocr2keepers.UpkeepIdentifier(p.Id.String())
+		checkBlockNumber, err := c.getCheckBlockNumberFromTxHash(p.TxHash, upkeepId)
+		if err != nil {
+			c.logger.Error("error while fetching checkBlockNumber from reorged report log: %w", err)
+			continue
+		}
+		l := ocr2keepers.PerformLog{
+			Key:             encoding.BasicEncoder{}.MakeUpkeepKey(checkBlockNumber, upkeepId),
+			TransmitBlock:   BlockKeyHelper[int64]{}.MakeBlockKey(p.BlockNumber),
+			TransactionHash: p.TxHash.Hex(),
+			Confirmations:   end - p.BlockNumber,
+		}
+		vals = append(vals, l)
+	}
+
+	return vals, nil
+}
+
+func (c *LogProvider) StaleReportLogs(ctx context.Context) ([]ocr2keepers.StaleReportLog, error) {
```

Comment:
nitpick, but I found this name confusing since it actually combines 3 log types and coerces them into one log type. Could we call this something generic like `AbortedReportLogs()` and make the return type generic? Not sure how annoying that would be. Not a strong opinion.